### PR TITLE
CTL locking script and CNFT minting policy

### PR DESCRIPTION
**Pull request by rynoV**

### CODE REVIEW

**Comment by samuelWilliams99** [comment]
Build step will likely fail, to be solved by #365 
Have verified it builds successfully locally

File: mlabs/efficient-nft-pab/Main.hs [hunk 1]
```diff
@@ -164,62 +156,25 @@ mkCollateral = do
...
-
-  -- in case of types changing, can help with request rewriting
-  -- Hask.putStr "seabug-mint-request: "
-  -- ByteString.putStrLn $
-  --   JSON.encode $
-  --     Mint (assetClass uCs $ fromString tokenName, mp)
-
-  -- Hask.putStr "unapplied-minting-policy: "
-  -- ByteString.putStrLn $ JSON.encode Token.policyDataScript
+  Hask.putStr "Unapplied sgNFT minting policy: "
```
Comment (rynoV) [inline mlabs/efficient-nft-pab/Main.hs:176]: The other one to log is the one from Lock.hs

Comment (samuelWilliams99) [reply mlabs/efficient-nft-pab/Main.hs:176]: Oh right, its not being logged at all right now?

Comment (rynoV) [reply mlabs/efficient-nft-pab/Main.hs:176]: Yeah it's not being logged, I logged it, saved it then removed the log. It was logged in the same was as the cnft minting policy

Comment (samuelWilliams99) [reply mlabs/efficient-nft-pab/Main.hs:177]: Have added now :)

File: mlabs/efficient-nft-pab/Main.hs [hunk 2]
```diff
@@ -176,6 +177,9 @@ main = do
   Hask.putStr "Unapplied sgNFT minting policy: "
   ByteString.putStrLn $ JSON.encode Token.policyDataScript
 
+  Hask.putStr "Unapplied locking validator: "
+  ByteString.putStrLn $ JSON.encode Lock.lockScriptUntyped
```
Comment (rynoV) [inline mlabs/efficient-nft-pab/Main.hs:181]: I'm not sure if `JSON.encode` gives the same result as `Text.decodeUtf8 . Base16.encode $ serialise`. I used the latter for the locking script saved in CTL right now

Comment (samuelWilliams99) [reply mlabs/efficient-nft-pab/Main.hs:181]: Looked into it, they both calldown to the Serialise typeclass, but unsure about the encoding, I'll switch to this method for all of them

**Comment by samuelWilliams99** [comment]
Verified builds on my machine

**Review by rynoV — COMMENTED** [review]
Looks good to me now

File: mlabs/efficient-nft-pab/Main.hs [hunk 3]
```diff
@@ -70,23 +69,16 @@ mkPolicy oref _ ctx =
...
-  --                      ]
+    $$(PlutusTx.compile [||(wrapMintingPolicy . mkPolicyUntyped)||])
+      `PlutusTx.applyCode` PlutusTx.liftCode (PlutusTx.toBuiltinData oref)
 
 policyUntypedScript :: Script
 policyUntypedScript =
   fromCompiledCode
-    $$(PlutusTx.compile [||(\a -> wrapMintingPolicy (mkPolicyUntyped a))||])
-
+    $$(PlutusTx.compile [||(wrapMintingPolicy . mkPolicyUntyped)||])
```
Comment (rynoV) [inline mlabs/efficient-nft-pab/Main.hs:82]: This actually slightly changes the encoded script from what I have been testing with, I'll make the update and test again just to be sure.

Comment (rynoV) [reply mlabs/efficient-nft-pab/Main.hs:82]: Tested as described in https://github.com/mlabs-haskell/seabug-contracts/issues/24, and also used it to successfully mint :+1:

**Pull request merged**